lib/Runtime/Library/InJavascript/Intl.js

```diff
@@ -1593,44 +1593,122 @@
+                }), 0);
+            }
+
+            {
+                // scope for the cache variable
+                let __date_prototype_toLocaleDateString_default_formatterStateObject_cache = undefined;
+                platform.registerBuiltInFunction(tagPublicFunction("Date.prototype.toLocaleDateString", function () {
+                    if (typeof this !== 'object' || !(this instanceof Date)) {
+                        platform.raiseNeedObjectOfType("Date.prototype.toLocaleDateString", "Date");
+                    }
+                    var value = callInstanceFunc(DateInstanceGetDate, new Date(this));
+                    if (isNaN(value) || !isFinite(value)) {
+                        return "Invalid Date";
+                    }
+
+                    let stateObject = undefined;
+
+                    let locale = undefined;
+                    let options = undefined;
+                    if (arguments.length >= 1) locale = arguments[0];
+                    if (arguments.length >= 2) options = arguments[1];
+
+                    if (!locale && !options) {
+                        // All default parameters (locale and options): this is the most valuable case to cache.
+                        if (__date_prototype_toLocaleDateString_default_formatterStateObject_cache) {
+                            stateObject = __date_prototype_toLocaleDateString_default_formatterStateObject_cache;
+                        } else {
+                            stateObject = setPrototype({}, null);
+                            InitializeDateTimeFormat(stateObject, undefined, ToDateTimeOptions(undefined, "any", "all"));
+                            __date_prototype_toLocaleDateString_default_formatterStateObject_cache = stateObject;
+                        }
+                    }
+
+                    if (!stateObject) {
+                        stateObject = setPrototype({}, null);
+                        InitializeDateTimeFormat(stateObject, locale, ToDateTimeOptions(options, "any", "all"));
+                    }
+
+                    return String(platform.formatDateTime(Internal.ToNumber(this), stateObject));
+                }), 1);
+            }
+
+            {
+                // scope for the cache variable
+                let __date_prototype_toLocaleTimeString_default_formatterStateObject_cache = undefined;
+                platform.registerBuiltInFunction(tagPublicFunction("Date.prototype.toLocaleTimeString", function () {
+                    if (typeof this !== 'object' || !(this instanceof Date)) {
+                        platform.raiseNeedObjectOfType("Date.prototype.toLocaleTimeString", "Date");
+                    }
+                    var value = callInstanceFunc(DateInstanceGetDate, new Date(this));
+                    if (isNaN(value) || !isFinite(value)) {
+                        return "Invalid Date";
+                    }
+
+                    let stateObject = undefined;
```

Comment:
refactor all these three instanced to one? this function is already so big to follow.

Reply:
well I get confused with other function - I take back that this function is big. But still good to refactor the code. All I see is different the cached var.

---
In reply to: [145771668](https://github.com/Microsoft/ChakraCore/pull/4006#discussion_r145771668) [](ancestors = 145771668)

Reply:
The call to instantiate the formatter takes different parameters. There's no way to pass a pointer in for the cache so assigning it would not propagate outside of the call. I'll see if there's something I can do here.

Reply:
I made a mistake in copy-paste. The parameters to the instantiation were supposed to be different (toLocaleString formats DateTime; toLocaleDateString, Date only; toLocaleTimeString, Time only).

Apparently this behavior was not covered by any test.

```
>eshost -h d8,ch-dev-che3,ch-master-latest -e "new Date().toLocaleString()"
#### d8
10/19/2017, 12:46:06 PM

#### ch-master-latest
‎10‎/‎19‎/‎2017‎ ‎12‎:‎46‎:‎06‎ ‎PM

#### ch-dev-che3
‎10‎/‎19‎/‎2017‎ ‎12‎:‎46‎:‎06‎ ‎PM

>eshost -h d8,ch-dev-che3,ch-master-latest -e "new Date().toLocaleTimeString()"
#### ch-master-latest
‎12‎:‎43‎:‎17‎ ‎PM

#### ch-dev-che3
‎10‎/‎19‎/‎2017‎ ‎12‎:‎43‎:‎17‎ ‎PM

#### d8
12:43:17 PM


>eshost -h d8,ch-dev-che3,ch-master-latest -e "new Date().toLocaleDateString()"
#### d8
10/19/2017

#### ch-dev-che3
‎10‎/‎19‎/‎2017‎ ‎12‎:‎43‎:‎38‎ ‎PM

#### ch-master-latest
‎10‎/‎19‎/‎2017
```

I will fix this and add tests to cover it.

Reply:
I figured out a refactoring that I think works well here.